Backport of:

From 773f7e8e3d16a0426c11edd7c3d8883ab6ee3a56 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@redhat.com>
Date: Mon, 13 Mar 2017 17:06:47 +0100
Subject: [PATCH] Introduced the %VERIFY_ALLOW_BROKEN priority string option

This allows enabling broken signature algorithms in certificate verification.

Signed-off-by: Nikos Mavrogiannopoulos <nmav@redhat.com>
---
 doc/cha-cert-auth.texi     | 1 +
 doc/cha-gtls-app.texi      | 4 ++++
 lib/priority.c             | 5 +++++
 lib/priority_options.gperf | 1 +
 4 files changed, 11 insertions(+)

--- a/doc/cha-cert-auth.texi
+++ b/doc/cha-cert-auth.texi
@@ -631,6 +631,7 @@ certificate chain, you can call
 @itemize
 @item @code{GNUTLS_VERIFY_ALLOW_SIGN_RSA_MD2}
 @item @code{GNUTLS_VERIFY_ALLOW_SIGN_RSA_MD5}
+@item @code{GNUTLS_VERIFY_ALLOW_BROKEN}
 @end itemize
 as in the following example:
 
--- a/doc/cha-gtls-app.texi
+++ b/doc/cha-gtls-app.texi
@@ -1369,6 +1369,10 @@ client hello.  Note that this should be
 try to reconnect with a downgraded protocol version. See RFC7507 for
 details.
 
+@item %VERIFY_ALLOW_BROKEN @tab
+will allow signatures with known to be broken algorithms (such as MD5 or
+SHA1) in certificate chains.
+
 @item %VERIFY_ALLOW_SIGN_RSA_MD5 @tab
 will allow RSA-MD5 signatures in certificate chains.
 
--- a/lib/priority.c
+++ b/lib/priority.c
@@ -861,6 +861,11 @@ static void enable_verify_allow_rsa_md5(
 	c->additional_verify_flags |=
 	    GNUTLS_VERIFY_ALLOW_SIGN_RSA_MD5;
 }
+static void enable_verify_allow_broken(gnutls_priority_t c)
+{
+	c->additional_verify_flags |=
+	    GNUTLS_VERIFY_ALLOW_BROKEN;
+}
 static void disable_crl_checks(gnutls_priority_t c)
 {
 	c->additional_verify_flags |=
--- a/lib/priority_options.gperf
+++ b/lib/priority_options.gperf
@@ -12,6 +12,7 @@ NO_TICKETS, enable_no_tickets
 NO_ETM, enable_no_etm
 NO_SESSION_HASH, enable_no_ext_master_secret
 STATELESS_COMPRESSION, enable_stateless_compression
+VERIFY_ALLOW_BROKEN, enable_verify_allow_broken
 VERIFY_ALLOW_SIGN_RSA_MD5, enable_verify_allow_rsa_md5
 VERIFY_DISABLE_CRL_CHECKS, disable_crl_checks
 SSL3_RECORD_VERSION, enable_ssl3_record_version
--- a/lib/priority_options.h
+++ b/lib/priority_options.h
@@ -1,6 +1,6 @@
 /* ANSI-C code produced by gperf version 3.0.4 */
 /* Command-line: gperf --global-table -t priority_options.gperf  */
-/* Computed positions: -k'1,$' */
+/* Computed positions: -k'9,$' */
 
 #if !((' ' == 32) && ('!' == 33) && ('"' == 34) && ('#' == 35) \
       && ('%' == 37) && ('&' == 38) && ('\'' == 39) && ('(' == 40) \
@@ -35,12 +35,12 @@ typedef void (*option_set_func)(gnutls_p
 #line 6 "priority_options.gperf"
 struct priority_options_st { const char *name; option_set_func func; };
 
-#define TOTAL_KEYWORDS 29
+#define TOTAL_KEYWORDS 30
 #define MIN_WORD_LENGTH 6
 #define MAX_WORD_LENGTH 32
 #define MIN_HASH_VALUE 6
-#define MAX_HASH_VALUE 52
-/* maximum key range = 47, duplicates = 0 */
+#define MAX_HASH_VALUE 74
+/* maximum key range = 69, duplicates = 0 */
 
 #ifdef __GNUC__
 __inline
@@ -54,104 +54,120 @@ hash (register const char *str, register
 {
   static const unsigned char asso_values[] =
     {
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      25, 53, 53,  5, 53, 53, 20, 53, 53, 53,
-      53, 53, 53, 53, 53,  5, 53, 10,  0, 25,
-      10,  0,  0, 53, 53, 10, 20,  0,  0, 53,
-      10, 53, 53,  0,  0,  0,  0, 30, 53,  5,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53, 53, 53, 53, 53,
-      53, 53, 53, 53, 53, 53
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      25, 75, 75, 15, 75, 75, 10, 75, 75, 75,
+      75, 75, 75, 75, 75,  5, 75, 75, 75, 10,
+      75, 25,  0, 10, 75,  0,  0, 30,  0, 10,
+      75, 75, 20,  0, 10,  5,  5,  0, 75,  0,
+      75, 75, 75, 75, 75,  0, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75, 75, 75, 75, 75,
+      75, 75, 75, 75, 75, 75
     };
-  return len + asso_values[(unsigned char)str[len - 1]] + asso_values[(unsigned char)str[0]];
+  register int hval = len;
+
+  switch (hval)
+    {
+      default:
+        hval += asso_values[(unsigned char)str[8]];
+      /*FALLTHROUGH*/
+      case 8:
+      case 7:
+      case 6:
+        break;
+    }
+  return hval + asso_values[(unsigned char)str[len - 1]];
 }
 
 static const struct priority_options_st wordlist[] =
   {
     {""}, {""}, {""}, {""}, {""}, {""},
-#line 12 "priority_options.gperf"
-    {"NO_ETM", enable_no_etm},
-    {""}, {""}, {""},
-#line 11 "priority_options.gperf"
-    {"NO_TICKETS", enable_no_tickets},
-#line 35 "priority_options.gperf"
-    {"NEW_PADDING", dummy_func},
-    {""},
+#line 9 "priority_options.gperf"
+    {"DUMBFW", enable_dumbfw},
+    {""}, {""}, {""}, {""},
+#line 29 "priority_options.gperf"
+    {"PROFILE_LOW", enable_profile_low},
+#line 32 "priority_options.gperf"
+    {"PROFILE_HIGH", enable_profile_high},
 #line 10 "priority_options.gperf"
     {"NO_EXTENSIONS", enable_no_extensions},
+#line 30 "priority_options.gperf"
+    {"PROFILE_LEGACY", enable_profile_legacy},
     {""},
-#line 13 "priority_options.gperf"
-    {"NO_SESSION_HASH", enable_no_ext_master_secret},
 #line 8 "priority_options.gperf"
     {"COMPAT", enable_compat},
-#line 25 "priority_options.gperf"
+#line 26 "priority_options.gperf"
     {"DISABLE_WILDCARDS", disable_wildcards},
-#line 21 "priority_options.gperf"
-    {"SAFE_RENEGOTIATION", enable_safe_renegotiation},
-#line 17 "priority_options.gperf"
-    {"SSL3_RECORD_VERSION", enable_ssl3_record_version},
-#line 20 "priority_options.gperf"
-    {"UNSAFE_RENEGOTIATION", enable_unsafe_renegotiation},
+#line 25 "priority_options.gperf"
+    {"FALLBACK_SCSV", enable_fallback_scsv},
+#line 15 "priority_options.gperf"
+    {"VERIFY_ALLOW_BROKEN", enable_verify_allow_broken},
+#line 11 "priority_options.gperf"
+    {"NO_TICKETS", enable_no_tickets},
 #line 14 "priority_options.gperf"
     {"STATELESS_COMPRESSION", enable_stateless_compression},
-#line 31 "priority_options.gperf"
-    {"PROFILE_HIGH", enable_profile_high},
+#line 28 "priority_options.gperf"
+    {"PROFILE_VERY_WEAK", enable_profile_very_weak},
+#line 33 "priority_options.gperf"
+    {"PROFILE_ULTRA", enable_profile_ultra},
+    {""},
+#line 13 "priority_options.gperf"
+    {"NO_SESSION_HASH", enable_no_ext_master_secret},
 #line 24 "priority_options.gperf"
-    {"FALLBACK_SCSV", enable_fallback_scsv},
-#line 30 "priority_options.gperf"
-    {"PROFILE_MEDIUM", enable_profile_medium},
-#line 16 "priority_options.gperf"
-    {"VERIFY_DISABLE_CRL_CHECKS", disable_crl_checks},
-#line 23 "priority_options.gperf"
     {"DISABLE_SAFE_RENEGOTIATION", disable_safe_renegotiation},
+#line 34 "priority_options.gperf"
+    {"PROFILE_SUITEB128", enable_profile_suiteb128},
+#line 22 "priority_options.gperf"
+    {"SAFE_RENEGOTIATION", enable_safe_renegotiation},
+#line 18 "priority_options.gperf"
+    {"SSL3_RECORD_VERSION", enable_ssl3_record_version},
+#line 21 "priority_options.gperf"
+    {"UNSAFE_RENEGOTIATION", enable_unsafe_renegotiation},
 #line 19 "priority_options.gperf"
+    {"LATEST_RECORD_VERSION", enable_latest_record_version},
+#line 37 "priority_options.gperf"
+    {"DEBUG_ALLOW_KEY_USAGE_VIOLATIONS", enable_server_key_usage_violations},
+    {""}, {""},
+#line 17 "priority_options.gperf"
+    {"VERIFY_DISABLE_CRL_CHECKS", disable_crl_checks},
+#line 12 "priority_options.gperf"
+    {"NO_ETM", enable_no_etm},
+#line 20 "priority_options.gperf"
     {"VERIFY_ALLOW_X509_V1_CA_CRT", dummy_func},
-#line 32 "priority_options.gperf"
-    {"PROFILE_ULTRA", enable_profile_ultra},
-#line 29 "priority_options.gperf"
-    {"PROFILE_LEGACY", enable_profile_legacy},
-#line 15 "priority_options.gperf"
+    {""}, {""},
+#line 16 "priority_options.gperf"
     {"VERIFY_ALLOW_SIGN_RSA_MD5", enable_verify_allow_rsa_md5},
-#line 22 "priority_options.gperf"
+#line 23 "priority_options.gperf"
     {"PARTIAL_RENEGOTIATION", enable_partial_safe_renegotiation},
-#line 36 "priority_options.gperf"
-    {"DEBUG_ALLOW_KEY_USAGE_VIOLATIONS", enable_server_key_usage_violations},
+#line 35 "priority_options.gperf"
+    {"PROFILE_SUITEB192", enable_profile_suiteb192},
     {""}, {""}, {""},
-#line 9 "priority_options.gperf"
-    {"DUMBFW", enable_dumbfw},
+#line 36 "priority_options.gperf"
+    {"NEW_PADDING", dummy_func},
 #line 27 "priority_options.gperf"
-    {"PROFILE_VERY_WEAK", enable_profile_very_weak},
-    {""}, {""}, {""},
-#line 18 "priority_options.gperf"
-    {"LATEST_RECORD_VERSION", enable_latest_record_version},
-#line 26 "priority_options.gperf"
     {"SERVER_PRECEDENCE", enable_server_precedence},
-    {""}, {""}, {""}, {""},
-#line 33 "priority_options.gperf"
-    {"PROFILE_SUITEB128", enable_profile_suiteb128},
-    {""}, {""}, {""},
-#line 28 "priority_options.gperf"
-    {"PROFILE_LOW", enable_profile_low},
-#line 34 "priority_options.gperf"
-    {"PROFILE_SUITEB192", enable_profile_suiteb192}
+    {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""},
+    {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""},
+    {""}, {""}, {""}, {""}, {""}, {""}, {""}, {""},
+#line 31 "priority_options.gperf"
+    {"PROFILE_MEDIUM", enable_profile_medium}
   };
 
 #ifdef __GNUC__
@@ -160,7 +176,7 @@ __inline
 __attribute__ ((__gnu_inline__))
 #endif
 #endif
-static const struct priority_options_st *
+const struct priority_options_st *
 in_word_set (register const char *str, register unsigned int len)
 {
   if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
